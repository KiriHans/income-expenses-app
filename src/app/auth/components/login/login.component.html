<div class="page-body-wrapper full-page-wrapper auth-page">
  <div class="content-wrapper d-flex align-items-center auth auth-bg-1 theme-one">
    <div class="row w-100">
      <div class="col-lg-4 mx-auto">
        <h2 class="text-center mb-4" style="color: white">Log in</h2>
        <div class="auto-form-wrapper">
          <form [formGroup]="loginForm" (ngSubmit)="loginUser()">
            <div class="form-group">
              <label class="label" for="email">Email</label>
              <div class="input-group">
                <input
                  id="email"
                  type="email"
                  class="form-control"
                  placeholder="Email"
                  formControlName="email"
                />
                <div class="input-group-append">
                  <span class="input-group-text">
                    <i class="fa fa-check-circle"></i>
                  </span>
                </div>
              </div>
            </div>
            <div class="form-group">
              <label class="label" for="password">Contrase√±a</label>
              <div class="input-group">
                <input
                  id="password"
                  type="password"
                  class="form-control"
                  placeholder="*********"
                  formControlName="password"
                />
                <div class="input-group-append">
                  <span class="input-group-text">
                    <i class="fa fa-check-circle"></i>
                  </span>
                </div>
              </div>
            </div>
            <div class="form-group">
              <button
                *ngIf="!isLoading()"
                [disabled]="loginForm.invalid"
                class="btn btn-primary submit-btn btn-block"
              >
                Login
              </button>
              <button
                *ngIf="isLoading()"
                [disabled]="true"
                type="button"
                class="btn btn-primary submit-btn btn-block"
              >
                <i class="fa fa-spin fa-sync"></i>
                Waiting...
              </button>
            </div>

            <div class="text-block text-center my-3">
              <span class="text-small font-weight-semibold">Are you not registered yet? </span>
              <br />
              <a routerLink="/auth/register" class="text-black text-small">Create new account</a>
            </div>
          </form>
        </div>
      </div>
    </div>
  </div>
  <!-- content-wrapper ends -->
</div>
