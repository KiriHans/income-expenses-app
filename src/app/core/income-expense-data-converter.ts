import {
  FirestoreDataConverter,
  QueryDocumentSnapshot,
  SnapshotOptions,
} from '@angular/fire/firestore';
import { IncomeExpense } from './models/income-expenses.model';

export const incomeExpenseConverter: FirestoreDataConverter<IncomeExpense> = {
  toFirestore: (incomeExpense: IncomeExpense) => {
    const { description, amount, type, uid } = incomeExpense;
    return { description, amount, type, uid };
  },
  fromFirestore: (
    snapshot: QueryDocumentSnapshot<IncomeExpense>,
    options: SnapshotOptions | undefined
  ) => {
    const data = snapshot.data(options);
    return new IncomeExpense(data.description, data.amount, data.type, data.uid);
  },
};
