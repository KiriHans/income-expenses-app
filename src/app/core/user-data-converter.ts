import {
  FirestoreDataConverter,
  QueryDocumentSnapshot,
  SnapshotOptions,
} from '@angular/fire/firestore';
import { UserApp } from './models/users.model';

export const userConverter: FirestoreDataConverter<UserApp> = {
  toFirestore: (user: UserApp) => {
    const { displayName, email } = user;
    return { displayName, email };
  },
  fromFirestore: (
    snapshot: QueryDocumentSnapshot<UserApp>,
    options: SnapshotOptions | undefined
  ) => {
    const data = snapshot.data(options);
    return new UserApp(data.id, data.displayName, data.email);
  },
};
