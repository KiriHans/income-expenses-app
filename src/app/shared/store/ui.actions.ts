import { createActionGroup, emptyProps } from '@ngrx/store';

export const UiActions = createActionGroup({
  source: 'UI',
  events: {
    'Load UI': emptyProps(),
    'Stop Load UI': emptyProps(),
  },
});
