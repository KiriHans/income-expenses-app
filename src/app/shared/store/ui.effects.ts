import { Injectable } from '@angular/core';
import { Actions, createEffect, ofType } from '@ngrx/effects';
import { map, concatMap } from 'rxjs/operators';
import { EMPTY } from 'rxjs';
import { UiActions } from './ui.actions';

@Injectable()
export class UiEffects {
  loadUis$ = createEffect(() => {
    return this.actions$.pipe(
      ofType(UiActions.loadUI),
      concatMap(() =>
        /** An EMPTY observable only emits completion. Replace with your own observable API request */
        EMPTY.pipe(map((data) => UiActions.loadUI()))
      )
    );
  });

  constructor(private actions$: Actions) {}
}
