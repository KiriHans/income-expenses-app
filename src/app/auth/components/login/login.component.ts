import { Component, OnInit, Signal } from '@angular/core';
import { CommonModule } from '@angular/common';
import { RouterLink } from '@angular/router';
import { FormBuilder, FormGroup, ReactiveFormsModule, Validators } from '@angular/forms';
import { Store } from '@ngrx/store';
import { LoginActions } from '../../../store/actions/auth.actions';
import { authFeature } from 'src/app/store/reducers/auth.reducer';

@Component({
  selector: 'app-login',
  standalone: true,
  imports: [CommonModule, RouterLink, ReactiveFormsModule],
  templateUrl: './login.component.html',
  styles: ['input { height: 100% }'],
})
export class LoginComponent implements OnInit {
  loginForm!: FormGroup;
  isLoading!: Signal<boolean>;

  constructor(private fb: FormBuilder, private store: Store) {}

  ngOnInit(): void {
    this.loginForm = this.fb.group({
      email: ['', [Validators.required, Validators.email]],
      password: ['', Validators.required],
    });

    this.isLoading = this.store.selectSignal(authFeature.selectLoading);
  }

  loginUser() {
    if (this.loginForm.invalid) return;

    this.store.dispatch(LoginActions.loginWithEmail(this.loginForm.value));
  }
}
