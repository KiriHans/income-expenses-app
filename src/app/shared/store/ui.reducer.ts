import { createFeature, createReducer, on } from '@ngrx/store';
import { UiActions } from './ui.actions';

export const uiFeatureKey = 'ui';

export interface UIState {
  isLoading: boolean;
}

export const initialState: UIState = {
  isLoading: false,
};

export const uiReducer = createReducer(
  initialState,
  on(UiActions.loadUI, (state): UIState => ({ ...state, isLoading: true })),
  on(UiActions.stopLoadUI, (state): UIState => ({ ...state, isLoading: false }))
);

export const uiFeature = createFeature({
  name: uiFeatureKey,
  reducer: uiReducer,
});
