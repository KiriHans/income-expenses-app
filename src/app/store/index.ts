import { isDevMode } from '@angular/core';
import { ActionReducerMap, MetaReducer } from '@ngrx/store';
import * as ui from './reducers/ui.reducer';
import * as auth from './reducers/auth.reducer';
import * as incomeExpense from './reducers/income-expense.reducer';

import { logoutMetaReducer } from './metareducers/logout.metareducer';
import { userStorageMetaReducer } from './metareducers/user-storage.metareducer';

export const stateFeatureKey = 'state';

export interface AppState {
  [ui.uiFeatureKey]: ui.UIState;
  [auth.authFeatureKey]: auth.State;
  [incomeExpense.incomeExpenseFeatureKey]?: incomeExpense.State;
}

export const reducers: ActionReducerMap<AppState> = {
  [ui.uiFeatureKey]: ui.uiFeature.reducer,
  [auth.authFeatureKey]: auth.userReducer,
};

export const metaReducersRoot: MetaReducer<AppState>[] = isDevMode()
  ? [logoutMetaReducer, userStorageMetaReducer]
  : [logoutMetaReducer, userStorageMetaReducer];
