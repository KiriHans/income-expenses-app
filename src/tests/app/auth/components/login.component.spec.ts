import { ComponentFixture, TestBed } from '@angular/core/testing';

import { RouterTestingModule } from '@angular/router/testing';
import { DebugElement } from '@angular/core';
import { By } from '@angular/platform-browser';

import { provideMockStore, MockStore } from '@ngrx/store/testing';
import { LoginComponent } from 'src/app/auth/components/login/login.component';
describe('LoginComponent', () => {
  let component: LoginComponent;
  let fixture: ComponentFixture<LoginComponent>;
  let el: DebugElement;
  let compiled: HTMLElement;

  let store: MockStore;
  const initialState = { user: { loading: false } };

  beforeEach(() => {
    TestBed.configureTestingModule({
      imports: [LoginComponent, RouterTestingModule],
      providers: [provideMockStore({ initialState })],
    });
    fixture = TestBed.createComponent(LoginComponent);
    component = fixture.componentInstance;
    el = fixture.debugElement;
    compiled = fixture.nativeElement;

    store = TestBed.inject(MockStore);

    fixture.detectChanges();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should match snapshot', () => {
    // expect(compiled).toMatchSnapshot();
  });

  it('should render a form with email and password inputs', () => {
    const logInTitle = el.query(By.css('h2'));
    const loginForm = el.query(By.css('form'));
    const [emailInput, passwordInput] = el.queryAll(By.css('.form-group>input'));
    console.log([emailInput, passwordInput]);
  });
});
