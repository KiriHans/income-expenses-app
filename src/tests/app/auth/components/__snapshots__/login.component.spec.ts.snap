// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginComponent should match snapshot 1`] = `
<div
  id="root1"
>
  <div
    class="page-body-wrapper full-page-wrapper auth-page"
  >
    <div
      class="content-wrapper d-flex align-items-center auth auth-bg-1 theme-one"
    >
      <div
        class="row w-100"
      >
        <div
          class="col-lg-4 mx-auto"
        >
          <h2
            class="text-center mb-4"
            style="color: white;"
          >
            Log in
          </h2>
          <div
            class="auto-form-wrapper"
          >
            <form
              class="ng-untouched ng-pristine ng-invalid"
              novalidate=""
            >
              <div
                class="form-group"
              >
                <label
                  class="label"
                  for="email"
                >
                  Email
                </label>
                <div
                  class="input-group"
                >
                  <input
                    class="form-control ng-untouched ng-pristine ng-invalid"
                    formcontrolname="email"
                    id="email"
                    placeholder="Email"
                    type="email"
                  />
                  <div
                    class="input-group-append"
                  >
                    <span
                      class="input-group-text"
                    >
                      <i
                        class="fa fa-check-circle"
                      />
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="form-group"
              >
                <label
                  class="label"
                  for="password"
                >
                  Contrase√±a
                </label>
                <div
                  class="input-group"
                >
                  <input
                    class="form-control ng-untouched ng-pristine ng-invalid"
                    formcontrolname="password"
                    id="password"
                    placeholder="*********"
                    type="password"
                  />
                  <div
                    class="input-group-append"
                  >
                    <span
                      class="input-group-text"
                    >
                      <i
                        class="fa fa-check-circle"
                      />
                    </span>
                  </div>
                </div>
              </div>
              <div
                class="form-group"
              >
                <button
                  class="btn btn-primary submit-btn btn-block"
                  disabled=""
                >
                   Login 
                </button>
                
                
              </div>
              <div
                class="text-block text-center my-3"
              >
                <span
                  class="text-small font-weight-semibold"
                >
                  Are you not registered yet? 
                </span>
                <br />
                <a
                  class="text-black text-small"
                  href="/auth/register"
                  routerlink="/auth/register"
                >
                  Create new account
                </a>
              </div>
            </form>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
